Adds ability to ignore parse exception console messages

dambrosio opened the pull request:
Allows a user to specify the `doxylink_parse_error_ignore_regexes` configuration value in the `Sphinx` _conf.py_ file to prevent `doxylink` tag parsing from printing specific messages to console. The value shall be a list of regular expressions that can be used to ignore specific errors reported from the parser.  Default is ``[]`` for backwards compatibility.

In this example, the first error message will no longer be logged:

```
doxylink_parse_error_ignore_regexes = [r"DEFINE.*"]

Skipping function company::DEFINE_bool(show, false, "Enable visualization"). Error reported from parser was: Expected ')', found ','  (at char 12), (line:1, col:13)
Skipping function transform::invR(transform_pb2.Rotation2f a). Error reported from parser was: Expected ')', found '.'  (at char 14), (line:1, col:15)
```

These parsing messages can overwhelm the console and mask users from seeing valuable error or warning messages provided by `Sphinx`

dambrosio committed:
feat: add ability to ignore parse exception console messages

Refs: #55

dambrosio commented on sphinxcontrib/doxylink/__init__.py:
```diff
@@ -1,10 +1,13 @@
     from .doxylink import setup_doxylink_roles
     app.add_config_value('doxylink', {}, 'env')
     app.add_config_value('doxylink_pdf_files', {}, 'env')
+    app.add_config_value(config.DOXYLINK_IGNORE_PARSE_EXCEPTIONS_KEY,
```
I will add this configuration to the _index.rst_ documentation.

dambrosio replied:
@JasperCraeghs the only reason I added that was so I didn't have to hardcode the configuration key for ignoring parse exceptions. I am totally cool with removing that and hardcoding per your recommendation.

JasperCraeghs reviewed (commented):
Thanks for this contribution. Instead of a boolean, we could support a list of regular expressions instead, for added granularity. What do you think?

This project has received contributions that we recently merged to the `master` branch but have not yet been released. These contributions will likely get rid of many of the parser warnings tnat you encounter. I will try to reach out to the owner of this repository again to make a release.

JasperCraeghs commented:
FYI We have released version [1.12.4](https://pypi.org/project/sphinxcontrib-doxylink/) this evening.

I triggered the unit tests in CI and some of them require an update. If you want any help, let me know. 🙂

dambrosio commented on sphinxcontrib/doxylink/config.py:
```diff
@@ -0,0 +1,5 @@
+DOXYLINK_KEY = 'doxylink'
+
+DOXYLINK_IGNORE_PARSE_EXCEPTIONS_KEY = (
+    '_'.join([DOXYLINK_KEY,'ignore_parse_exceptions'])
```
That is my bad. Do we have a default linter/formatter for Python in this repo, to minimize these lint issues?

JasperCraeghs replied:
No, not yet

dambrosio commented:
> FYI We have released version [1.12.4](https://pypi.org/project/sphinxcontrib-doxylink/) this evening.
> 
> I triggered the unit tests in CI and some of them require an update. If you want any help, let me know. 🙂

I can test this out to see if it fixes some of the parse exceptions in our code base.

dambrosio commented:
> Thanks for this contribution. Instead of a boolean, we could support a list of regular expressions instead, for added granularity. What do you think?

I would be down to change this behavior. Are you thinking a single regex or multiple regexs? Would the idea be to match specific text inside of the `ParseException` error message?

ie.
```
Skipping function company::DEFINE_string(logs_directory, "", "Path to the directory."). Error reported from parser was: Expected ')', found ','  (at char 15), (line:1, col:16)
```

So we could add a regex that would be something like `company::DEFINE` to only match specific `member_kind`, `member_symbol` or `arglist` values?

dambrosio commented:
> > FYI We have released version [1.12.4](https://pypi.org/project/sphinxcontrib-doxylink/) this evening.
> > I triggered the unit tests in CI and some of them require an update. If you want any help, let me know. 🙂
> 
> I can test this out to see if it fixes some of the parse exceptions in our code base.

I still see parse exceptions in our code base. I think some are probably due to our usage of `gflags` as well as selectively including specific source for documentation. In either case, I still think having a mechanism to disable these prints would be helpful from our end.

dambrosio committed:
docs: Added configuration documenation for doxylink_parse_exception_ignore_regex_list

Refs: #55

dambrosio committed:
test: ensure tests pass with a None parse_exception_ignore_pattern

Refs: #55

JasperCraeghs committed:
Ignore tag file generated by test suite

JasperCraeghs committed:
Use report_warning instead of print; don't combine regexes into a single pattern for robustness; rename config variable

JasperCraeghs committed:
revert type annotations for create_role as it's complex

JasperCraeghs commented:
- Combining the regex patterns with `|` without grouping (with parentheses) may lead to unexpected results, I think. I chose to evaluate each regex one by one. Python stores the compiled patterns in cache.
- I added a test case.
- I renamed  the configuration variable `doxylink_parse_exception_ignore_regex_list` to `doxylink_parse_error_ignore_regexes`. I'll update the PR description to reflect our recent changes.

Please review my contribution.

JasperCraeghs committed:
Test that parser error, a proper Sphinx warning, is ignored due to doxylink_parse_error_ignore_regexes

JasperCraeghs committed:
Address deprecation warning about whitelist_externals

dambrosio commented:
> * I replaced `print` with `report_warning` so that it becomes a proper Sphinx warning that can be converted into an error with `sphinx-build -W`
> * Combining the regex patterns with `|` without grouping (with parentheses) may lead to unexpected results, I think. I chose to evaluate each regex one by one. Python stores the compiled patterns in cache.
> * I added a test case.
> * I renamed  the configuration variable `doxylink_parse_exception_ignore_regex_list` to `doxylink_parse_error_ignore_regexes`. I'll update the PR description to reflect our recent changes.
> 
> Please review my contribution.

Thanks for updating this @JasperCraeghs, your changes look great :tada:!

When this merges do you have a plan for what the release timeline looks like?

Cheers!

JasperCraeghs commented:
Released as [1.13.0 ](https://github.com/sphinx-contrib/doxylink/releases/tag/1.13.0)

kaycebasques commented:
Should this docs section be updated to mention the new config option? https://sphinxcontrib-doxylink.readthedocs.io/en/stable/#configuration-values

Presumably you can't link to symbols that Doxylink has ignored, is that correct?

kaycebasques commented:
I was looking at the wrong docs. Still not seeing the new conf option in `latest`, though: https://sphinxcontrib-doxylink.readthedocs.io/en/latest/

dambrosio commented:
> I was looking at the wrong docs. Still not seeing the new conf option in `latest`, though: https://sphinxcontrib-doxylink.readthedocs.io/en/latest/

Looks like the `latest` ReadtheDocs has not been built for some time: https://app.readthedocs.org/projects/sphinxcontrib-doxylink/builds/.

The config option is documented in source under _doc/index.rst_.

@JasperCraeghs does this seem correct?
